dm/syncer/validate_worker.go

```diff
@@ -0,0 +1,348 @@
+}
+
+func (vw *validateWorker) validateDeletedRows(cond *Cond) ([]string, error) {
+	targetRows, err := vw.getTargetRows(cond)
+	if err != nil {
+		return []string{}, err
+	}
+
+	failedRows := make([]string, 0, len(targetRows))
+	for key := range targetRows {
+		failedRows = append(failedRows, key)
+	}
+	return failedRows, nil
+}
+
+func (vw *validateWorker) validateInsertAndUpdateRows(rows []*rowChange, cond *Cond) ([]string, error) {
+	var failedRows []string
+	sourceRows := getSourceRowsForCompare(rows)
+	targetRows, err := vw.getTargetRows(cond)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(targetRows) > len(sourceRows) {
+		// if this happens, downstream should have removed the primary key
+		vw.L.Debug("more data on downstream, may come from other client")
+	}
+
+	tableInfo := cond.Table.Info
+	for key, sourceRow := range sourceRows {
+		targetRow, ok := targetRows[key]
+		if !ok {
+			failedRows = append(failedRows, key)
+			continue
+		}
+
+		eq, err := vw.compareData(sourceRow, targetRow, tableInfo.Columns[:cond.ColumnCnt])
+		if err != nil {
+			return nil, err
+		}
+		if !eq {
+			failedRows = append(failedRows, key)
+		}
+	}
+	return failedRows, nil
+}
+
+// a simplified version of https://github.com/pingcap/tidb-tools/blob/d9fdfa2f9040aab3fab7cd11774a82226f467fe7/sync_diff_inspector/utils/utils.go#L487-L606
+func (vw *validateWorker) compareData(sourceData, targetData []*sql.NullString, columns []*model.ColumnInfo) (bool, error) {
+	for i, column := range columns {
+		data1, data2 := sourceData[i], targetData[i]
+		if data1.Valid != data2.Valid {
+			return false, nil
+		}
+		str1, str2 := data1.String, data2.String
+		if str1 == str2 {
+			continue
+		} else if column.FieldType.Tp == tidbmysql.TypeFloat || column.FieldType.Tp == tidbmysql.TypeDouble {
+			// source and target data have different precision?
+			num1, err1 := strconv.ParseFloat(str1, 64)
+			num2, err2 := strconv.ParseFloat(str2, 64)
+			if err1 != nil || err2 != nil {
+				// should not happen
+				return false, errors.Errorf("convert %s, %s to float failed, err1: %v, err2: %v", str1, str2, err1, err2)
+			}
+			if math.Abs(num1-num2) <= 1e-6 {
+				continue
+			}
+		}
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (vw *validateWorker) getTargetRows(cond *Cond) (map[string][]*sql.NullString, error) {
+	tctx := tcontext.NewContext(vw.ctx, log.L())
+	fullTableName := cond.Table.Target.String()
+	columnNames := make([]string, 0, cond.ColumnCnt)
+	for i := 0; i < cond.ColumnCnt; i++ {
+		col := cond.Table.Info.Columns[i]
+		columnNames = append(columnNames, dbutil.ColumnName(col.Name.O))
+	}
+	columns := strings.Join(columnNames, ", ")
+	rowsQuery := fmt.Sprintf("SELECT /*!40001 SQL_NO_CACHE */ %s FROM %s WHERE %s",
+		columns, fullTableName, cond.GetWhere())
+	rows, err := vw.conn.QuerySQL(tctx, rowsQuery, cond.GetArgs()...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	result := make(map[string][]*sql.NullString)
```

Comment:
Similarly, if we scan through and fetch all rows (rather than iterate them one by one) in big data situation, will it be a bit risky.

Reply:
yes, add a todo here

https://github.com/buchuitoudegou/tiflow/blob/ebba8bb19d0083a8653b815617ff55cea4a50fe8/dm/syncer/validate_worker.go#L134-L145

Reply:
Making the worker being driven by a fixed number of rows in addition to polling the channel every once in awhile might solve this (i.e. restrict the number of rows) and be more intuitive.

Reply:
you mean the way we do in prototype?

in that case we're keeping 2 pending row changes set, one is current accumulating(with a max batch size), another is what's left from previous validation which needs retry and its size may exceed the batch size. it's more complex, and we cannot merge row changes to eliminate duplicates, also for the second one we'll face the same problem as in `todo`.

Reply:
It is too complicated the way we did in prototype. I think handling the situation that accumulating too many rows is plausible at the time we receive the row change from the validator (according to the newly added todo).

Reply:
add in https://github.com/pingcap/tiflow/pull/4767/commits/36f1486309c578f90405fab8242f56250c42ae55